File: database_integration/enhance_db/src/bin/20_create_individuals_countries.rs
/// Create individuals_countries table:
/// Associate each individual with a modern country based on:
/// 1. nationality (first priority)
/// 2. deathplace (second priority)
/// 3. birthplace (third priority)
/// Includes an "origins" column indicating the source.
use anyhow::Result;
use indicatif::{ProgressBar, ProgressStyle};
use rusqlite::{params, Connection};
use std::collections::HashMap;
use std::fs;
use std::io::Write;

const DB_PATH: &str = "data/humans_clean.sqlite3";
const TASK_LOG: &str = "task.log";
const BATCH_SIZE: usize = 50_000;

fn log(msg: &str) {
    println!("{}", msg);
    let mut f = fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(TASK_LOG)
        .unwrap();
    writeln!(f, "{}", msg).unwrap();
}

fn main() -> Result<()> {
    log("=== Step 20: Create individuals_countries table ===");

    let conn = Connection::open(DB_PATH)?;
    conn.execute_batch(
        "PRAGMA journal_mode=WAL; PRAGMA synchronous=NORMAL; PRAGMA cache_size=-2000000;",
    )?;

    // Build nationality name -> (country_name, iso_a3_code) lookup
    log("[20] Building nationality lookup...");
    let mut nat_lookup: HashMap<String, (String, String)> = HashMap::new();
    {
        let mut stmt = conn.prepare(
            "SELECT name_en, country_name, iso_a3_code FROM nationalities WHERE country_name IS NOT NULL AND iso_a3_code IS NOT NULL"
        )?;
        let rows = stmt.query_map([], |r| {
            Ok((
                r.get::<_, String>(0)?,
                r.get::<_, String>(1)?,
                r.get::<_, String>(2)?,
            ))
        })?;
        for r in rows {
            let (name, country, iso) = r?;
            nat_lookup.insert(name, (country, iso));
        }
    }
    log(&format!("[20] Nationality lookup: {} entries", nat_lookup.len()));

    // Build city name -> (country_name, iso_a3) lookup
    log("[20] Building city lookup...");
    let mut city_lookup: HashMap<String, (String, String)> = HashMap::new();
    {
        let mut stmt = conn.prepare(
            "SELECT name_en, country_name, iso_a3 FROM cities WHERE country_name IS NOT NULL AND iso_a3 IS NOT NULL"
        )?;
        let rows = stmt.query_map([], |r| {
            Ok((
                r.get::<_, String>(0)?,
                r.get::<_, String>(1)?,
                r.get::<_, String>(2)?,
            ))
        })?;
        for r in rows {
            let (name, country, iso) = r?;
            // For duplicate city names, first one wins (they usually have same country)
            city_lookup.entry(name).or_insert((country, iso));
        }
    }
    log(&format!("[20] City lookup: {} entries", city_lookup.len()));

    // Drop and create the target table
    log("[20] Creating individuals_countries table...");
    conn.execute_batch("DROP TABLE IF EXISTS individuals_countries;")?;
    conn.execute_batch(
        "CREATE TABLE individuals_countries (
            wikidata_id TEXT PRIMARY KEY,
            country_name TEXT NOT NULL,
            iso_a3_code TEXT NOT NULL,
            origins TEXT NOT NULL
        );"
    )?;

    // Count total individuals
    let total: i64 = conn.query_row("SELECT COUNT(*) FROM individuals", [], |r| r.get(0))?;
    log(&format!("[20] Total individuals: {}", total));

    // Process in batches using LIMIT/OFFSET via rowid ordering
    let pb = ProgressBar::new(total as u64);
    pb.set_style(
        ProgressStyle::default_bar()
            .template("{msg} [{bar:40}] {pos}/{len} ({eta})")
            .unwrap(),
    );
    pb.set_message("Processing individuals");

    let mut offset: i64 = 0;
    let mut matched_nationality = 0u64;
    let mut matched_death = 0u64;
    let mut matched_birth = 0u64;
    let mut unmatched = 0u64;
    let mut total_inserted = 0u64;

    loop {
        // Read a batch
        let mut batch: Vec<(String, Option<String>, Option<String>, Option<String>)> = Vec::with_capacity(BATCH_SIZE);
        {
            let mut stmt = conn.prepare_cached(
                "SELECT wikidata_id, nationalities_en, deathcity_en, birthcity_en
                 FROM individuals
                 ORDER BY rowid
                 LIMIT ?1 OFFSET ?2"
            )?;
            let rows = stmt.query_map(params![BATCH_SIZE as i64, offset], |r| {
                Ok((
                    r.get::<_, String>(0)?,
                    r.get::<_, Option<String>>(1)?,
                    r.get::<_, Option<String>>(2)?,
                    r.get::<_, Option<String>>(3)?,
                ))
            })?;
            for r in rows {
                batch.push(r?);
            }
        }

        if batch.is_empty() {
            break;
        }

        // Process batch and insert
        conn.execute_batch("BEGIN TRANSACTION;")?;
        {
            let mut insert = conn.prepare_cached(
                "INSERT OR IGNORE INTO individuals_countries (wikidata_id, country_name, iso_a3_code, origins)
                 VALUES (?1, ?2, ?3, ?4)"
            )?;

            for (wikidata_id, nationalities_en, deathcity_en, birthcity_en) in &batch {
                let mut found = false;

                // Priority 1: nationality
                if let Some(nats) = nationalities_en {
                    for nat_name in nats.split("; ") {
                        let nat_name = nat_name.trim();
                        if let Some((country, iso)) = nat_lookup.get(nat_name) {
                            insert.execute(params![wikidata_id, country, iso, "nationality"])?;
                            matched_nationality += 1;
                            total_inserted += 1;
                            found = true;
                            break;
                        }
                    }
                }

                if found {
                    continue;
                }

                // Priority 2: deathplace
                if let Some(city) = deathcity_en {
                    let city = city.trim();
                    if let Some((country, iso)) = city_lookup.get(city) {
                        insert.execute(params![wikidata_id, country, iso, "deathplace"])?;
                        matched_death += 1;
                        total_inserted += 1;
                        found = true;
                    }
                }

                if found {
                    continue;
                }

                // Priority 3: birthplace
                if let Some(city) = birthcity_en {
                    let city = city.trim();
                    if let Some((country, iso)) = city_lookup.get(city) {
                        insert.execute(params![wikidata_id, country, iso, "birthplace"])?;
                        matched_birth += 1;
                        total_inserted += 1;
                        found = true;
                    }
                }

                if !found {
                    unmatched += 1;
                }
            }
        }
        conn.execute_batch("COMMIT;")?;

        pb.inc(batch.len() as u64);
        offset += batch.len() as i64;

        // Log progress every 500k
        if offset % 500_000 < BATCH_SIZE as i64 {
            log(&format!(
                "[20] Progress: {}/{} processed, {} inserted (nat:{}, death:{}, birth:{}), {} unmatched",
                offset, total, total_inserted, matched_nationality, matched_death, matched_birth, unmatched
            ));
        }
    }
    pb.finish();

    // Create indexes
    log("[20] Creating indexes...");
    conn.execute_batch("CREATE INDEX IF NOT EXISTS idx_indcountries_country ON individuals_countries(country_name);")?;
    conn.execute_batch("CREATE INDEX IF NOT EXISTS idx_indcountries_iso ON individuals_countries(iso_a3_code);")?;
    conn.execute_batch("CREATE INDEX IF NOT EXISTS idx_indcountries_origins ON individuals_countries(origins);")?;

    // Final stats
    let final_count: i64 = conn.query_row(
        "SELECT COUNT(*) FROM individuals_countries",
        [],
        |r| r.get(0),
    )?;

    log(&format!("[20] === Final Statistics ==="));
    log(&format!("[20] Total individuals: {}", total));
    log(&format!("[20] Total matched: {}", total_inserted));
    log(&format!("[20]   via nationality: {}", matched_nationality));
    log(&format!("[20]   via deathplace: {}", matched_death));
    log(&format!("[20]   via birthplace: {}", matched_birth));
    log(&format!("[20] Unmatched: {}", unmatched));
    log(&format!("[20] Rows in individuals_countries: {}", final_count));

    // Show top countries
    let mut top = conn.prepare(
        "SELECT country_name, iso_a3_code, COUNT(*) as cnt FROM individuals_countries GROUP BY country_name ORDER BY cnt DESC LIMIT 15"
    )?;
    let rows: Vec<(String, String, i64)> = top
        .query_map([], |r| Ok((r.get(0)?, r.get(1)?, r.get(2)?)))?
        .filter_map(|r| r.ok())
        .collect();
    log("[20] Top 15 countries:");
    for (name, iso, cnt) in &rows {
        log(&format!("[20]   {} ({}) -> {}", name, iso, cnt));
    }

    log("=== Step 20 complete ===");
    Ok(())
}

File: scripts/database_integration_scripts_V2/legacy/20_create_individuals_countries.py
"""20 — Create `individuals_countries` from a 3-priority match
(nationality -> deathplace -> birthplace).

Mirrors `enhance_db/src/bin/20_create_individuals_countries.rs`.

  Inputs : individuals (wikidata_id, nationalities_en, deathcity_en, birthcity_en)
           nationalities (name_en, country_name, iso_a3_code)
           cities (name_en, country_name, iso_a3)
  Output : individuals_countries (wikidata_id PK, country_name, iso_a3_code, origins)
"""

from __future__ import annotations

import sqlite3
import tempfile
from pathlib import Path

from common import (
    DB_PATH,
    insert_rows,
    log,
    open_db,
    parse_run_mode,
    transaction,
)

BATCH_SIZE = 50_000


def run(conn: sqlite3.Connection) -> None:
    log("[DB] 20: Create individuals_countries...")

    nat_lookup: dict[str, tuple[str, str]] = {}
    for name, country, iso in conn.execute(
        "SELECT name_en, country_name, iso_a3_code FROM nationalities "
        "WHERE country_name IS NOT NULL AND iso_a3_code IS NOT NULL"
    ):
        nat_lookup[name] = (country, iso)
    log(f"[20] Nationality lookup: {len(nat_lookup)} entries")

    city_lookup: dict[str, tuple[str, str]] = {}
    for name, country, iso in conn.execute(
        "SELECT name_en, country_name, iso_a3 FROM cities "
        "WHERE country_name IS NOT NULL AND iso_a3 IS NOT NULL"
    ):
        if name not in city_lookup:
            city_lookup[name] = (country, iso)
    log(f"[20] City lookup: {len(city_lookup)} entries")

    conn.execute("DROP TABLE IF EXISTS individuals_countries")
    conn.execute(
        """
        CREATE TABLE individuals_countries (
            wikidata_id TEXT PRIMARY KEY,
            country_name TEXT NOT NULL,
            iso_a3_code TEXT NOT NULL,
            origins TEXT NOT NULL
        )
        """
    )

    total = conn.execute("SELECT COUNT(*) FROM individuals").fetchone()[0]
    log(f"[20] Total individuals: {total}")

    cur = conn.execute(
        "SELECT wikidata_id, nationalities_en, deathcity_en, birthcity_en FROM individuals"
    )
    try:
        from tqdm import tqdm
        iterator = tqdm(cur, total=total, desc="ind_countries", unit="row")
    except ImportError:
        iterator = cur

    matched_nat = matched_death = matched_birth = unmatched = 0
    insert_sql = (
        "INSERT OR IGNORE INTO individuals_countries "
        "(wikidata_id, country_name, iso_a3_code, origins) VALUES (?, ?, ?, ?)"
    )
    buf: list[tuple] = []

    with transaction(conn):
        ins = conn.cursor()
        for wid, nats, death, birth in iterator:
            row = None
            if nats:
                for n in nats.split("; "):
                    hit = nat_lookup.get(n.strip())
                    if hit:
                        row = (wid, hit[0], hit[1], "nationality")
                        matched_nat += 1
                        break
            if row is None and death:
                hit = city_lookup.get(death.strip())
                if hit:
                    row = (wid, hit[0], hit[1], "deathplace")
                    matched_death += 1
            if row is None and birth:
                hit = city_lookup.get(birth.strip())
                if hit:
                    row = (wid, hit[0], hit[1], "birthplace")
                    matched_birth += 1

            if row is None:
                unmatched += 1
            else:
                buf.append(row)
                if len(buf) >= BATCH_SIZE:
                    ins.executemany(insert_sql, buf)
                    buf.clear()
        if buf:
            ins.executemany(insert_sql, buf)

    conn.execute("CREATE INDEX IF NOT EXISTS idx_indcountries_country ON individuals_countries(country_name)")
    conn.execute("CREATE INDEX IF NOT EXISTS idx_indcountries_iso ON individuals_countries(iso_a3_code)")
    conn.execute("CREATE INDEX IF NOT EXISTS idx_indcountries_origins ON individuals_countries(origins)")
    conn.commit()

    final = conn.execute("SELECT COUNT(*) FROM individuals_countries").fetchone()[0]
    log(f"[20] Inserted {final} (nat:{matched_nat} death:{matched_death} birth:{matched_birth} unmatched:{unmatched})")


def _sample_main() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        db = Path(tmp) / "sample.sqlite3"
        with sqlite3.connect(db) as seed:
            seed.execute(
                "CREATE TABLE individuals (wikidata_id TEXT PRIMARY KEY, name_en TEXT, "
                "nationalities_en TEXT, deathcity_en TEXT, birthcity_en TEXT)"
            )
            seed.execute(
                "CREATE TABLE nationalities (wikidata_id TEXT, name_en TEXT, "
                "country_name TEXT, iso_a3_code TEXT)"
            )
            seed.execute(
                "CREATE TABLE cities (id TEXT, name_en TEXT, country_name TEXT, iso_a3 TEXT)"
            )
            insert_rows(seed, "nationalities", [
                {"wikidata_id": "Q142", "name_en": "French", "country_name": "France", "iso_a3_code": "FRA"},
                {"wikidata_id": "Q30", "name_en": "American", "country_name": "United States", "iso_a3_code": "USA"},
            ])
            insert_rows(seed, "cities", [
                {"id": "Q90", "name_en": "Paris", "country_name": "France", "iso_a3": "FRA"},
                {"id": "Q60", "name_en": "New York", "country_name": "United States", "iso_a3": "USA"},
            ])
            insert_rows(seed, "individuals", [
                {"wikidata_id": "P1", "name_en": "Hugo", "nationalities_en": "French",
                 "deathcity_en": None, "birthcity_en": None},
                {"wikidata_id": "P2", "name_en": "Smith", "nationalities_en": None,
                 "deathcity_en": "New York", "birthcity_en": None},
                {"wikidata_id": "P3", "name_en": "Doe", "nationalities_en": None,
                 "deathcity_en": None, "birthcity_en": "Paris"},
                {"wikidata_id": "P4", "name_en": "Nobody", "nationalities_en": None,
                 "deathcity_en": None, "birthcity_en": None},
            ])

        with open_db(db) as conn:
            run(conn)
            for row in conn.execute(
                "SELECT wikidata_id, country_name, iso_a3_code, origins FROM individuals_countries ORDER BY wikidata_id"
            ):
                log(f"  {row}")


if __name__ == "__main__":
    if parse_run_mode() == "full":
        with open_db(DB_PATH) as conn:
            run(conn)
    else:
        _sample_main()
